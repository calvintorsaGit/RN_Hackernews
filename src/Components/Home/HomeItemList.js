import React from 'react';
import {StyleSheet, View, Text, TouchableOpacity} from 'react-native';
import {navigateToComments} from '../../Utils/Navigation'
import get from 'lodash/get'
import {useNavigation} from "@react-navigation/native";

const topSection = (props) => (
    <View style={styles.textContainer}>
        <Text style={styles.titleText}>{props.item.title}</Text>
    </View>
)

const bottomSection = (props) => {
    const navigation = useNavigation();
    const totalComments = get(props, 'item.kids', [])
    return <View style={styles.SectionContainer}>
        <Text>{props.item.score} points by {props.item.by} | </Text>
        <TouchableOpacity onPress={navigateToComments.bind(this, navigation, props.item)}>
            <Text style={styles.commentText}>{totalComments.length} comments</Text>
        </TouchableOpacity>
    </View>
}

const _renderItemView = (props) => (
    <View style={styles.parentContainer}>
        <Text style={styles.numberText}>{props.index}.</Text>
        <View>
            {topSection(props)}
            {bottomSection(props)}
        </View>
    </View>
)

const HomeItemList = (props) => (
    <View style={styles.viewContainer}>
        {_renderItemView(props)}
    </View>
)

const styles = StyleSheet.create({
    viewContainer : {
        marginVertical: 4,
        marginHorizontal: 8
    },
    parentContainer: {
        flexDirection: 'row',
        paddingRight: 32
    },
    textContainer:{
        flex:1
    },
    SectionContainer: {
        marginVertical: 4,
        flexDirection: 'row'
    },
    titleText:{
      fontWeight: 'bold',
      fontSize: 16
    },
    numberText: {
        marginRight: 4,
        fontSize: 20,
        fontWeight: 'bold'
    },
    commentText: {
        color: '#FF8C00',
        fontWeight: 'bold'
    },
});

export default HomeItemList;
