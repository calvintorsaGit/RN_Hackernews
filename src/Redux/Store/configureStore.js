import { createStore, combineReducers, applyMiddleware } from 'redux';
import ItemListReducer from '../Reducer/ItemListReducer';
import logger from 'redux-logger';
import thunk from 'redux-thunk';

const rootReducer = combineReducers(
    { itemListReducer: ItemListReducer}
);
const configureStore = () => {
    return createStore(rootReducer,applyMiddleware(thunk, logger));
}
export default configureStore;
