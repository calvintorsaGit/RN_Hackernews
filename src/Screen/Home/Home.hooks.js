import {useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {
    fetchIdStories,
    fetchMoreStories,
} from '../../Redux/Action/ItemListAction';

import {actions} from '../../Redux/Reducer/ItemListReducer'

const homeHooks = () => {
    const [refreshing, setRefreshing] = useState(false);
    const [reload, setReload] = useState(true);
    const [listIndex, setListIndex] = useState(20);
    // const [idStories, setIdStories] = useState([]);
    const itemListReducer = useSelector(state => state.itemListReducer);
    const idStories = itemListReducer.idStories;
    const stories = itemListReducer.stories;

    const dispatch = useDispatch();
    const refreshStories = () => dispatch(actions.clearRefreshStories())
    const getIdStories = () => dispatch(fetchIdStories('topstories'));
    const getMoreStories = (ids) => dispatch(fetchMoreStories(ids, 'topstories'));

    return {
        stories,
        refreshing,
        setRefreshing,
        listIndex,
        setListIndex,
        idStories,
        getIdStories,
        getMoreStories,
        refreshStories,
        reload,
        setReload
    };
};

export default homeHooks;
