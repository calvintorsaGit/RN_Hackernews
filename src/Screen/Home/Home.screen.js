import React, {useEffect} from 'react';
import {SafeAreaView, StyleSheet, FlatList} from 'react-native';
import HomeItemList from '../../Components/Home/HomeItemList'
import homeHooks from "./Home.hooks";

const initateData = (hooks) => {
    useEffect(() => {
        hooks.getIdStories();
    }, [hooks.idStories]);

    if (hooks.idStories.length > 0 && hooks.reload) {
        hooks.setReload(false);
        const selectedStory = hooks.idStories.slice(0, 15);
        hooks.getMoreStories(selectedStory)
        hooks.setRefreshing(false);
    }
}

const _renderItem = ({item, index}) => {
    return <HomeItemList item={item} index={index + 1}/>
}

export const _onRefresh = (hooks) => {
    hooks.setRefreshing(true)
    hooks.setReload(true);
    hooks.refreshStories();
}

export const _onLoadMore = (hooks) => {
    const selectedStory = hooks.idStories.slice(hooks.listIndex + 1, hooks.listIndex + 10);
    hooks.setListIndex(hooks.listIndex + 10);
    hooks.getMoreStories(selectedStory)
}

const Home = () => {
    const hooks = homeHooks()
    initateData(hooks)

    return (
        <SafeAreaView style={styles.backgroundStyle}>
            <FlatList
                data={hooks.stories}
                renderItem={_renderItem}
                onRefresh={() => _onRefresh(hooks)}
                refreshing={hooks.refreshing}
                onEndReachedThreshold={0.4}
                onEndReached={() => _onLoadMore(hooks)}
            />
        </SafeAreaView>
    );
};

const styles = StyleSheet.create({
    backgroundStyle: {
        flex: 1,
        backgroundColor: '#f8f4ec',
    }
});

export default Home
