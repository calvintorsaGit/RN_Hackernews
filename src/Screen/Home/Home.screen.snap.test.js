import React from 'react';
import renderer from 'react-test-renderer';
import Home from './Home.screen';
import { Provider } from "react-redux"
import fetch from "node-fetch";
import configureStore from "../../Redux/Store/configureStore";
const store = configureStore()
globalThis.fetch = fetch

const HomeWrapper = () => (
    <Provider store={store}>
        <Home/>
    </Provider>
);

test('renders correctly', () => {
    const tree = renderer.create(<HomeWrapper />).toJSON();
    expect(tree).toMatchSnapshot();
});
