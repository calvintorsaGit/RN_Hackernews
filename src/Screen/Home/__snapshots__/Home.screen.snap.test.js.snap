// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f8f4ec",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.4}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
